File: examples/python/get_latest_trade.py
"""
Example: Fetch the latest trades from Pump.fun

This script demonstrates how to use the PumpFunAPI client to fetch the latest trades
across all tokens on Pump.fun.
"""

import json
import sys
import logging
from datetime import datetime
from pathlib import Path

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler('pumpfun_api.log')
    ]
)
logger = logging.getLogger(__name__)

# Add the parent directory to the path so we can import the client
sys.path.append(str(Path(__file__).parent.parent.parent))
try:
    from utils.python_client import PumpFunAPI
except ImportError as e:
    logger.error("Failed to import PumpFunAPI. Make sure you've installed the required dependencies.")
    logger.error(f"Error: {e}")
    sys.exit(1)

def format_timestamp(timestamp):
    """Convert Unix timestamp to human-readable format."""
    try:
        return datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
    except (TypeError, ValueError):
        return str(timestamp)

def main():
    logger.info("Starting Pump.fun API example: Get Latest Trades")
    
    try:
        # Initialize the API client with debug logging
        logger.debug("Initializing PumpFunAPI client")
        client = PumpFunAPI()
        
        # Fetch the latest trades
        logger.info("Fetching the latest trades...")
        result = client.get_latest_trades(limit=5)
        
        # Debug: Log the raw API response
        logger.debug(f"Raw API response: {json.dumps(result, indent=2)}")
        
        # Handle different response formats
        if isinstance(result, dict):
            if 'trades' in result and result['trades']:
                print("\nLatest Trades:")
                print("-" * 100)
                print(f"{'Token':<15} {'Amount In':<15} {'Amount Out':<15} {'Type':<8} {'Timestamp':<20} {'Tx Hash'}")
                print("-" * 100)
                
                for trade in result['trades']:
                    print(
                        f"{trade.get('symbol', 'N/A'):<15} "
                        f"{trade.get('sol_amount', 'N/A'):<15} "
                        f"{trade.get('token_amount', 'N/A'):<15} "
                        f"{'BUY' if trade.get('is_buy') else 'SELL':<8} "
                        f"{format_timestamp(trade.get('timestamp')):<20} "
                        f"{trade.get('signature', 'N/A')}"
                    )
            else:
                # Handle direct trade object (non-array response)
                print("\nLatest Trade:")
                print("-" * 100)
                print(json.dumps(result, indent=2, default=str))
                
                # Log the trade details
                logger.info(f"Trade: {result.get('symbol')} - "
                           f"SOL: {result.get('sol_amount')}, "
                           f"Tokens: {result.get('token_amount')}, "
                           f"Type: {'BUY' if result.get('is_buy') else 'SELL'}")
        else:
            logger.warning(f"Unexpected response format: {type(result)}")
            print("\nUnexpected response format. Raw response:")
            print(json.dumps(result, indent=2, default=str))
            
    except Exception as e:
        logger.error(f"An error occurred: {str(e)}", exc_info=True)
        print(f"\nError: {str(e)}")
        print("Check pumpfun_api.log for more details.")
        sys.exit(1)
    
    logger.info("Script completed successfully")
    print("\nDone! Check pumpfun_api.log for detailed logs.")

if __name__ == "__main__":
    main()

File: examples/javascript/get_latest_trade.js
/**
 * Pump.fun API Example: Get Latest Market Data
 * 
 * This example demonstrates how to fetch and display the latest market data
 * from the Pump.fun API with enhanced formatting and error handling.
 */

const { PumpFunAPI } = require('../../utils/js_client');

/**
 * Format a trade object for display
 * @param {Object} trade - Trade data
 * @param {PumpFunAPI} client - API client instance for formatting
 * @returns {string} Formatted trade string
 */
function formatTrade(trade, client) {
  if (!trade) return 'No trade data available';
  
  const timestamp = trade.timestamp ? new Date(trade.timestamp * 1000).toLocaleString() : 'N/A';
  const solAmount = client._formatSol(trade.sol_amount || 0);
  const tokenAmount = client._formatNumber(trade.token_amount || 0);
  const price = trade.price ? client._formatNumber(trade.price) : 'N/A';
  
  return [
    `🔄 ${trade.is_buy ? 'BUY' : 'SELL'}`,
    `Token: ${trade.symbol || 'N/A'} (${trade.mint || 'N/A'})`,
    `Amount: ${tokenAmount} ${trade.symbol || 'tokens'}`,
    `Value: ${solAmount}`,
    `Price: ${price} SOL`,
    `User: ${trade.user ? trade.user.substring(0, 6) + '...' + trade.user.slice(-4) : 'N/A'}`,
    `Time: ${timestamp}`,
    `Tx: ${trade.signature ? trade.signature.substring(0, 10) + '...' + trade.signature.slice(-8) : 'N/A'}`,
    trade.image_uri ? `Image: ${trade.image_uri}` : ''
  ].filter(Boolean).join('\n  ');
}

/**
 * Format a coin object for display
 * @param {Object} coin - Coin data
 * @param {PumpFunAPI} client - API client instance for formatting
 * @returns {string} Formatted coin string
 */
function formatCoin(coin, client) {
  if (!coin) return 'No coin data available';
  
  const marketCap = coin.market_cap ? `$${client._formatNumber(coin.market_cap, 2)}` : 'N/A';
  const price = coin.price ? `$${client._formatNumber(coin.price, 6)}` : 'N/A';
  const created = coin.created_at ? new Date(coin.created_at * 1000).toLocaleString() : 'N/A';
  
  return [
    `💰 ${coin.name || 'Unnamed Token'} (${coin.symbol || 'N/A'})`,
    `Price: ${price}`,
    `Market Cap: ${marketCap}`,
    `Mint: ${coin.mint || 'N/A'}`,
    `Created: ${created}`,
    coin.description ? `Description: ${coin.description.substring(0, 100)}${coin.description.length > 100 ? '...' : ''}` : '',
    coin.image_uri ? `Image: ${coin.image_uri}` : ''
  ].filter(Boolean).join('\n  ');
}

/**
 * Main function to fetch and display market data
 */
async function getMarketData() {
  // Initialize the client with debug mode enabled
  const client = new PumpFunAPI({
    // apiKey: 'your-api-key', // Optional: Add your API key for authenticated requests
  });
  client.debug = true; // Enable debug logging

  console.log('📊 Pump.fun Market Data');
  console.log('='.repeat(40));
  
  try {
    // 1. Get latest trades
    console.log('\n🔍 Fetching latest trades...');
    try {
      const response = await client._request('GET', '/trades/latest', { limit: 3 });
      
      if (response) {
        // Handle single trade object
        if (response.mint) {
          console.log('\n🔄 Latest Trade:');
          console.log('='.repeat(40));
          console.log(formatTrade(response, client));
        } 
        // Handle array of trades (if the API changes in the future)
        else if (Array.isArray(response) && response.length > 0) {
          console.log(`\n🔄 Latest ${response.length} Trades:`);
          console.log('='.repeat(40));
          response.forEach((trade, index) => {
            console.log(`\nTrade #${index + 1}:`);
            console.log(formatTrade(trade, client));
          });
        } else {
          console.log('ℹ️  No recent trades found');
          console.log('Response structure:', JSON.stringify(response, null, 2));
        }
      } else {
        console.log('❌ No response received for trades');
      }
    } catch (error) {
      console.error('\n❌ Error fetching trades:', error.message);
      if (error.response) {
        console.error('Status:', error.response.status);
        console.error('Details:', error.response.data?.error || 'No additional details');
      }
    }

    // 2. Get latest coins
    console.log('\n🔍 Fetching latest coins...');
    try {
      const response = await client._request('GET', '/coins/latest', { limit: 3, includeNsfw: false });
      
      if (response) {
        // Handle single coin object
        if (response.mint) {
          console.log('\n🆕 Latest Coin:');
          console.log('='.repeat(40));
          console.log(formatCoin(response, client));
        } 
        // Handle array of coins (if the API changes in the future)
        else if (Array.isArray(response) && response.length > 0) {
          console.log(`\n🆕 Latest ${response.length} Coins:`);
          console.log('='.repeat(40));
          response.forEach((coin, index) => {
            console.log(`\nCoin #${index + 1}:`);
            console.log(formatCoin(coin, client));
          });
        } else {
          console.log('ℹ️  No coins found');
          console.log('Response structure:', JSON.stringify(response, null, 2));
        }
      } else {
        console.log('❌ No response received for coins');
      }
    } catch (error) {
      console.error('\n❌ Error fetching coins:', error.message);
      if (error.response) {
        console.error('Status:', error.response.status);
        console.error('Details:', error.response.data?.error || 'No additional details');
      }
    }

  } catch (error) {
    console.error('\n❌ Unhandled error:', error.message);
    if (error.stack) {
      console.error('Stack trace:', error.stack.split('\n').slice(0, 3).join('\n') + '...');
    }
  }
  
  console.log('\n' + '='.repeat(40));
  console.log('✅ Data fetch complete');
}

// Run the example
console.log('Starting Pump.fun API example...\n');
getMarketData().catch(console.error);
